docs/proposals/20231122-security-groups.md

```diff
@@ -0,0 +1,225 @@
+  - "@jichenjc"
+  - "@mdbooth"
+creation-date: 2023-11-22
+last-updated: 2023-11-22
+
+status: provisional
+---
+
+## Title
+
+Flexible managed security groups
+
+## Table of Contents
+
+- [Flexible managed security groups](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+- CNI - Container Network Interface (CNI) is a specification for configuring network interfaces for Linux containers.
+  CNI defines the specification of a REST API for network drivers to configure network interfaces in Linux containers.
+  CNI also defines a specification of the configuration file format for the network drivers to plug into container runtime environments.
+- Controller - A controller is a control loop that watches the shared state of the cluster through the API server and makes changes
+  attempting to move the current state towards the desired state. Examples of controllers in CAPO are OpenStackCluster and OpenStackMachine.
+- Egress - Egress traffic is traffic that is leaving the cluster.
+- Ingress - Ingress traffic is traffic that is coming into the cluster.
+- Security Group - A security group is a named set of rules that control traffic to and from virtual machine instances.
+  The rules can specify which ports and protocols to allow, and which IP address ranges to allow traffic to or from.
+  Security groups can be associated with multiple instances, and multiple security groups can be associated with a single instance.
+  Security groups act as a firewall for associated instances, controlling both inbound and outbound traffic at the instance level.
+
+## Summary
+
+This proposal moves the management of security groups into a dedicated controller and introduces a new API to manage more security groups and rules.
+
+## Motivation
+
+We want users to be able to not think about the security groups that are needed for their Control Plane and Worker Nodes. However, we want to be able to provide a way for users to have more control over their security groups and not have to worry about rules that are not needed by their cluster.
+
+### Goals
+
+1. Move the management of security groups into a dedicated controller since we will have to reconcile the assignement of security groups to the machines.
```

Comment:
The question is really whether it makes sense for us. CAPA has a multitude of security groups already: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/3373251e75afd25e11b256a645a95469b0139328/api/v1beta2/network_types.go#L399-L418

If they don't need to alter their security group attachments they don't need a separate controller.

Reply:
Something interesting [here](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/1256#issuecomment-1829999292) which could justify a dedicated CRD & controller, eventually.
I'm waiting for @mkjpryor's feedback as well.

Reply:
I've pushed https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1774 which change a bit the logic. I hope you'll fine it convenient. A new controller is not required and not easy to implement because managed security groups are managed in the cluster and applied to the machine.